region/src/main/java/org/motechproject/nms/region/service/impl/LanguageLocationCodesImportServiceImpl.java

```diff
@@ -0,0 +1,219 @@
+
+        if (null != state && null != district) {
+            importRecordForStateAndDistrict(languageLocationCode, language, circle, defaultForCircle, state, district);
+        } else if (null != state) {
+            importRecordForState(languageLocationCode, language, circle, defaultForCircle, state);
+        } else if (null != district) {
+            importRecordForDistrict(languageLocationCode, language, circle, defaultForCircle, district);
+        } else {
+            throw new CsvImportDataException("Both state and district are null");
+        }
+    }
+
+    private void importRecordForStateAndDistrict(String languageLocationCode, Language language, Circle circle,
+                                                 Boolean defaultForCircle, State state, District district) {
+        verify(Objects.equals(district.getState().getCode(), state.getCode()),
+                "District's state does not match the state supplied in the CSV record");
+        verifyStateInCircle(circle, state);
+        verifyDistrictLanguageLocation(district);
+        addDistrictToLanguageLocation(languageLocationCode, language, circle, defaultForCircle, Arrays.asList(district));
+    }
+
+    private void importRecordForState(String languageLocationCode, Language language, Circle circle,
+                                      Boolean defaultForCircle, State state) {
+        verifyStateInCircle(circle, state);
+        for (District stateDistrict : state.getDistricts()) {
+            verifyDistrictLanguageLocation(stateDistrict);
+        }
+        addDistrictToLanguageLocation(languageLocationCode, language, circle, defaultForCircle, state.getDistricts());
+    }
+
+    private void importRecordForDistrict(String languageLocationCode, Language language, Circle circle,
+                                         Boolean defaultForCircle, District district) {
+        LOGGER.warn("State not specified for the '{}' district", district.getName());
+        verifyStateInCircle(circle, district.getState());
+        verifyDistrictLanguageLocation(district);
+        addDistrictToLanguageLocation(languageLocationCode, language, circle, defaultForCircle, Arrays.asList(district));
+    }
+
+    private void addDistrictToLanguageLocation(String languageLocationCode, Language language, Circle circle, Boolean defaultForCircle, Collection<District> districts) {
+        LanguageLocation languageLocation = languageLocationDataService.findByCode(languageLocationCode);
+        if (null != languageLocation) {
+            verify(Objects.equals(languageLocation.getLanguage().getName(), language.getName()),
+                    "Language location exists, yet its language does not match the language supplied in the CSV record");
+            verify(Objects.equals(languageLocation.getCircle().getName(), circle.getName()),
+                    "Language location exists, yet its circle does not match the circle supplied in the CSV record");
```

Comment:
This will have to change to a contains check or be removed once the relation is switched to many to many